fix(core): Sanitize tool parameters to fix 400 API errors

BigUncle opened the pull request:
> **Note**
> This PR depends on the test stabilization fix in **#3296**. Please review and merge that PR first. This PR includes the same test fix temporarily to pass CI checks, and it will be removed by rebasing onto `main` once #3296 is merged.

## TLDR

This pull request provides a robust fix for a persistent 400 API error caused by improper tool parameter schemas being sent to the Gemini API. It directly resolves issue #2141 and several duplicates by sanitizing tool function declarations before they are sent, ensuring compatibility with the API's requirements.

**Key Fixes:**

- **Sanitizes `format` property:** For `STRING` type parameters, it removes any `format` value that is not `enum` or `date-time`, which was the primary cause of the 400 errors.
- **Improves Tool Registry Robustness:** Adds `Array.isArray` checks to ensure `function_declarations` are correctly structured, which likely prevents the `missing field` errors also reported.

**This PR closes:** #2141, #2706, #3178, #2237
**This PR likely also resolves:** #2522

## Dive Deeper

The root cause of the `Persistent 400 API error` was that the tool discovery mechanism would generate function declarations with `format` properties (e.g., `uri`, `email`) that are not supported by the Gemini API for `STRING` parameters. The API strictly requires that the `format` for a string be either `enum` or `date-time`.

This PR introduces a `sanitizeParameters` function that recursively traverses the tool schema. It inspects every `STRING` parameter and sets its `format` to `undefined` if it's not one of the allowed values. This sanitization happens right before the tool is registered, ensuring that any request sent to the Gemini API contains a valid tool definition.

Additionally, issue #2141 and the related #2522 also reported errors like `...items: missing field.`. This often happens when an expected array structure is malformed. The secondary fix in this PR, which validates that `function_declarations` and `functionDeclarations` are indeed arrays before processing them, adds a layer of type safety that should prevent these malformed definitions from causing API errors.

## Reviewer Test Plan

1.  Pull down this branch.
2.  Run `npm run preflight`. All tests should pass consistently.
3.  To manually verify, one could set up a local tool discovery command that produces a tool with an unsupported string format (e.g., `"format": "uri"`).
    - **Before this PR:** Running a prompt that uses this tool would result in a 400 API error.
    - **After this PR:** The command should execute successfully without any API errors, as the unsupported format is sanitized before the request is made.

## Testing Matrix

|          | Local (WSL2) | CI  | Notes                                                            |
| -------- | :----------: | :-: | :--------------------------------------------------------------- |
| npm run  |      ✅      | ✅  | All checks pass. The rebase onto `upstream/main` was successful. |
| npx      |     N/A      | N/A | Not applicable for this change.                                  |
| Docker   |     N/A      | N/A | Not applicable for this change.                                  |
| Podman   |      -       |  -  | Not applicable for this change.                                  |
| Seatbelt |      -       |  -  | Not applicable for this change.                                  |

## Linked issues / bugs

- **Depends on #3296**
- Fixes #2141
- Fixes #2706
- Fixes #3178
- Fixes #2237
- Likely resolves #2522

BigUncle commented on packages/core/src/tools/tool-registry.ts:
```diff
@@ -160,27 +160,49 @@ export class ToolRegistry {
+              func.description ?? '',
+              (func.parameters ?? {}) as Record<string, unknown>,
+            ),
+          );
```
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

BigUncle commented:
/gemini review

NTaylorMullen commented on packages/core/src/tools/tool-registry.ts:
```diff
@@ -160,27 +162,121 @@ export class ToolRegistry {
-          functions.push(...tool['functionDeclarations']);
-        } else if (tool['name']) {
-          functions.push(tool);
+      try {
```
Could you extract this block into a specific function to resolve tools for readability?

BigUncle replied:
I have extracted it into a new function called `discoverAndRegisterToolsFromCommand`, thank you!

NTaylorMullen reviewed (commented):
@BigUncle we're super close! Thank you for the updates. Left a few more comments

BigUncle commented:
> @BigUncle we're super close! Thank you for the updates. Left a few more comments

Hi @NTaylorMullen,

Thank you so much for your detailed feedback! I've addressed all your points in the latest commit:

*   **`shell-quote` dependency**: You were right, I've moved the `shell-quote` dependency to `packages/core/package.json`. Thanks for catching that!
*   **`useLoadingIndicator.test.ts`**: The code removal was indeed unintentional. I've restored it in the latest version.
*   **Redundant check in `tool-registry.ts`**: I've removed the extra check as you suggested, since `parse` already handles it.
*   **Readability of `tool-registry.ts`**: I've extracted the tool resolution logic into a new `discoverAndRegisterToolsFromCommand` function to improve readability.

I really appreciate you taking the time to review my code. Your suggestions have made it much better!

BigUncle commented on package.json:
```diff
@@ -87,5 +88,7 @@
   },
-  "dependencies": {}
+  "dependencies": {
+    "shell-quote": "^1.8.3"
```
I've moved the `shell-quote` dependency to `packages/core/package.json`. Thank you!

BigUncle commented on packages/cli/src/ui/hooks/useLoadingIndicator.test.ts:
```diff
@@ -41,14 +41,12 @@ describe('useLoadingIndicator', () => {
     expect(WITTY_LOADING_PHRASES).toContain(
       result.current.currentLoadingPhrase,
     );
-    const initialPhrase = result.current.currentLoadingPhrase;
```
I made a mistake in the operation. I have restored it.

BigUncle commented:
> Thank you for all the back and forth and the contribution @BigUncle !

Thank you so much, @NTaylorMullen!

I really appreciate you taking the time for such a detailed review and for all the guidance. A special thanks for helping get this over the finish line by resolving the final merge conflict.

As this is my first PR to be merged in a major open-source project, I'm absolutely thrilled. I learned a ton from the process.

Looking forward to contributing more in the future!
